### PR TITLE
Denys/feature/image upload

### PR DESCRIPTION
**Pull request by DenisBurimov**

### CODE REVIEW

File: app/views/users.py [hunk 1]
```diff
@@ -112,10 +111,10 @@ def delete(id: int):
     return "ok", 200
 
 
-@bp.route("/<user_unique_id>", methods=["GET"])
+@bp.route("/profile", methods=["GET"])
 @login_required
-def user_profile(user_unique_id: str):
-    user_query = m.User.select().where(m.User.unique_id == user_unique_id)
+def user_profile():
+    user_query = m.User.select().where(m.User.unique_id == current_user.unique_id)
```
Comment (kipod) [inline app/views/users.py:117]:
```python
user = current_user

File: app/views/users.py [hunk 2]
```diff
@@ -126,4 +125,86 @@ def user_profile(user_unique_id: str):
...
+    user: m.User | None = db.session.scalar(query)
+
+    if not user:
+        log(log.INFO, "User not found")
+        flash("Incorrect reset password link", "danger")
+        return redirect(url_for("main.index"))
+
+    image_upload(user)
+
+    return {"upload logo status": "ok"}, 200
```
Comment (kipod) [inline app/views/users.py:150]:
```python
@bp.route("/logo-upload", methods=["GET", "POST"])
@login_required
def logo_upload():
    image_upload(current_user)
    return {}, 200

File: app/views/users.py [hunk 3]
```diff
@@ -126,4 +125,86 @@ def user_profile(user_unique_id: str):
...
+
+
+@bp.route("/edit_email")
+def edit_email():
```
Comment (kipod) [inline app/views/users.py:154]: login_required
and same as function before

File: app/views/users.py [hunk 4]
```diff
@@ -126,4 +125,86 @@ def user_profile(user_unique_id: str):
...
+
+    if not user:
+        log(log.INFO, "User not found")
+        flash("Incorrect reset password link", "danger")
+        return redirect(url_for("main.index"))
+    return render_template("user/email_edit.html", user=user)
+
+
+@bp.route("/save_email", methods=["GET", "POST"])
+def save_email():
```
Comment (kipod) [inline app/views/users.py:166]: login_required
and same as function before

File: app/views/users.py [hunk 5]
```diff
@@ -126,4 +125,86 @@ def user_profile(user_unique_id: str):
...
+    form = f.EmailEditForm()
+    if form.validate_on_submit():
+        user.email = form.email.data
+        user.save()
+
+    return render_template("user/email_save.html", user=user)
+
+
+@bp.route("/edit_phone")
+def edit_phone():
```
Comment (kipod) [inline app/views/users.py:184]: login_required
and same as function before

File: app/views/users.py [hunk 6]
```diff
@@ -126,4 +125,86 @@ def user_profile(user_unique_id: str):
...
+
+    if not user:
+        log(log.INFO, "User not found")
+        flash("Incorrect reset password link", "danger")
+        return redirect(url_for("main.index"))
+    return render_template("user/phone_edit.html", user=user)
+
+
+@bp.route("/save_phone", methods=["GET", "POST"])
+def save_phone():
```
Comment (kipod) [inline app/views/users.py:196]: same

File: test_flask/conftest.py [hunk 7]
```diff
@@ -14,6 +14,7 @@ def app():
     app.config.update(
         {
             "TESTING": True,
+            "DEBUG": True,
```
Comment (kipod) [inline test_flask/conftest.py:17]: a you really need it?

File: test_flask/db/__init__.py [hunk 8]
```diff
@@ -88,9 +91,21 @@ def generate_test_users(num_objects: int = NUM_TEST_USERS):
 
 
 def generate_test_events(num_objects: int = NUM_TEST_EVENTS):
+    folder_path = "test_flask/locations_pictures"  # replace with your folder path
```
Comment (kipod) [inline test_flask/db/__init__.py:94]: folder_path -> FOLDER_PATH
this is constant

File: test_flask/test_admin.py [hunk 9]
```diff
@@ -0,0 +1,66 @@
+import os
+from PIL import Image
+from io import BytesIO
+from flask.testing import FlaskClient
+from test_flask.utils import login
+from app import models as m
+from werkzeug.datastructures import FileStorage
```
Comment (kipod) [inline test_flask/test_admin.py:7]: format imports

**Review by kipod — CHANGES_REQUESTED** [review]
Good work.
Please work on my comments

File: app/controllers/image_upload.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        except Exception as e:
+            log(log.ERROR, "Error saving image: [%s]", e)
+            flash("Error saving image", "danger")
+            return redirect(url_for("auth.image_upload", user_unique_id=user.unique_id))
+
+        try:
+            picture = m.Picture(
+                filename=file.filename.split("/")[-1],
+                file=img_byte_arr,
+                mimetype=file.content_type,
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/controllers/image_upload.py:39]: Removed

File: app/controllers/image_upload.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
+import io
+from PIL import Image
+from flask import request, redirect, url_for, flash, current_app as app
+from app import models as m
+from app.logger import log
+
+
+def image_upload(user: m.User):
+    if request.method == "POST":
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/controllers/image_upload.py:9]: Done

File: app/controllers/image_upload.py [hunk 12]
```diff
@@ -0,0 +1,48 @@
+import io
+from PIL import Image
+from flask import request, redirect, url_for, flash, current_app as app
+from app import models as m
+from app.logger import log
+
+
+def image_upload(user: m.User):
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/controllers/image_upload.py:8]: Done

File: app/controllers/image_upload.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+                file=img_byte_arr,
+                mimetype=file.content_type,
+            ).save()
+            user.picture_id = picture.id
+            user.save()
+            flash("Logo uploaded", "success")
+        except Exception as e:
+            log(log.ERROR, "Error saving image: [%s]", e)
+            flash("Error saving image", "danger")
+            return redirect(url_for("main.index"))
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/controllers/image_upload.py:47]: Fixed. Now it is a completely controllers logic, that is used to save images

File: app/models/location.py [hunk 14]
```diff
@@ -36,11 +41,15 @@ class Location(db.Model, ModelMixin):
         unique=True,
         nullable=False,
     )
-    # TODO: add image
+    picture_id: orm.Mapped[int] = orm.mapped_column(
+        sa.Integer, sa.ForeignKey("pictures.id"), nullable=True
+    )
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/models/location.py:46]: Thanks for the idea

File: app/models/payment.py [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+
+class Payment(db.Model, ModelMixin):
+    __tablename__ = "payments"
+
+    id: orm.Mapped[int] = orm.mapped_column(
+        sa.Integer,
+        primary_key=True,
+        nullable=False,
+    )
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/models/payment.py:22]: DONE

File: app/models/payment.py [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    buyer_id: orm.Mapped[int] = orm.mapped_column(
+        sa.Integer,
+        sa.ForeignKey("users.id"),
+        nullable=False,
+    )
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/models/payment.py:27]: Good point. Done

File: app/models/payment.py [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+        sa.ForeignKey("events.id"),
+        nullable=False,
+    )
+    ticket_id: orm.Mapped[int] = orm.mapped_column(
+        sa.Integer,
+        sa.ForeignKey("tickets.id"),
+        nullable=False,
+    )
+    description: orm.Mapped[str] = orm.mapped_column(
+        sa.String(255),
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/models/payment.py:39]: Done

File: app/models/user.py [hunk 18]
```diff
@@ -57,12 +60,24 @@ class User(db.Model, UserMixin, ModelMixin):
         unique=True,
         nullable=False,
     )
-    image: orm.Mapped[bytes] = orm.mapped_column(sa.LargeBinary, nullable=True)
+    picture_id: orm.Mapped[int] = orm.mapped_column(
+        sa.Integer, sa.ForeignKey("pictures.id"), nullable=True
+    )
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/models/user.py:65]: Done

File: test_flask/test_admin.py (hunk 9 above)
Comment (DenisBurimov) [reply test_flask/test_admin.py:7]: Done

File: app/templates/user/email_edit.html [hunk 19]
```diff
@@ -0,0 +1,7 @@
+<form method="post" id="profile_email" hx-post="{{ url_for('user.save_email') }}" hx-target="#profile_email"
+    hx-swap="innerHTML">
+    <input name="email" type="email" id="email_input" type="text" value="{{user.email}}">
+    <button type="submit" class="bg-blue-300 text-white py-1 px-2 rounded-md">
+        save
+    </button>
+</form>
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/templates/user/email_edit.html:7]: Autoformatting removes it in html

File: app/templates/user/profile.html [hunk 20]
```diff
@@ -1,4 +1,52 @@
...
+        Payments:
+        {{ payments }}
+    </div>
+
+</div>
+
+{% endblock %}
+{% block scripts %}
+<script src="{{ url_for('static', filename='js/user.js') }}" type="text/javascript" defer></script>
 {% endblock %}
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/templates/user/profile.html:67]: Prettier removes EOLs

File: app/templates/user/phone_edit.html [hunk 21]
```diff
@@ -0,0 +1,7 @@
+<form method="post" id="profile_phone" hx-post="{{ url_for('user.save_phone') }}" hx-target="#profile_phone"
+    hx-swap="innerHTML">
+    <input name="phone" type="phone" id="phone_input" type="text" value="{{user.phone}}">
+    <button type="submit" class="bg-blue-300 text-white py-1 px-2 rounded-md">
+        save
+    </button>
+</form>
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/templates/user/phone_edit.html:7]: Prettier removes them

File: app/controllers/jinja_globals.py [hunk 22]
```diff
@@ -22,3 +24,13 @@ def time_delta(created_at: datetime):
 
 def cut_seconds(created_at: datetime):
     return created_at.strftime("%Y-%m-%d %H:%M")
+
+
+def stream_picture(picture_id):
+    picture_query = m.Picture.select().where(m.Picture.id == picture_id)
+    picture = db.session.scalar(picture_query)
+    if not picture:
+        return None
```
[the remark this answers is not in the document]
Comment (DenisBurimov) [reply app/controllers/jinja_globals.py:33]: Removed this global function to use property that returns base64 source

**Pull request merged**